meteor/client/ui/App.tsx

```diff
@@ -374,9 +365,162 @@ export const App = translateWithTracker(() => {
 						</ErrorBoundary>
 					</div>
 				</Router>
-			)
-		}
+			</UserSubscriptionReadyContext.Provider>
+		</UserContext.Provider>
+	)
+}
+
+/**
+ * This is a poly-fill, replicating the behavior of a <Navigate> element in React-Router-DOM v6.
+ * TODO: Use React-Router-DOM element once React-Router-DOM is upgraded to v6.
+ */
+const Navigate = React.memo(function Navigate({ to }: { to: string }): null {
+	const history = useHistory()
+
+	useEffect(() => {
+		history.push(to)
+	}, [to])
+
+	return null
+})
+
+const RequireAuth = React.memo(function RequireAuth({ children }: React.PropsWithChildren<{}>) {
+	const user = useContext(UserContext)
+	const ready = useContext(UserSubscriptionReadyContext)
+
+	let isAuthenticated = false
+	let isWorking = false
+	if (!Settings.enableUserAccounts) {
+		isAuthenticated = true
+		isWorking = false
+	} else if (!ready) {
+		isAuthenticated = false
+		isWorking = true
+	} else if (ready && user) {
+		isAuthenticated = true
+		isWorking = false
 	}
-)
+
+	if (isWorking) return <Spinner />
+	if (isAuthenticated) return <>{children}</>
+
+	return <Navigate to="/" />
+})
+
+function useRoles(user: User | null, subsReady: boolean) {
+	const location = window.location
+
+	const [roles, setRoles] = useState(
+		Settings.enableUserAccounts
+			? {
+					studio: false,
+					configure: false,
+					developer: false,
+					testing: false,
+					service: false,
+			  }
+			: {
+					studio: getAllowStudio(),
+					configure: getAllowConfigure(),
+					developer: getAllowDeveloper(),
+					testing: getAllowTesting(),
+					service: getAllowService(),
+			  }
+	)
+
+	useEffect(() => {
+		if (!Settings.enableUserAccounts) {
+			if (!location.search) return
+
+			const params = queryStringParse(location.search)
+
+			if (params['studio']) setAllowStudio(params['studio'] === '1')
+			if (params['configure']) setAllowConfigure(params['configure'] === '1')
+			if (params['develop']) setAllowDeveloper(params['develop'] === '1')
+			if (params['testing']) setAllowTesting(params['testing'] === '1')
+			if (params['service']) setAllowService(params['service'] === '1')
+
+			if (params['admin']) {
+				const val = params['admin'] === '1'
+				setAllowStudio(val)
+				setAllowConfigure(val)
+				setAllowDeveloper(val)
+				setAllowTesting(val)
+				setAllowService(val)
+			}
+
+			setRoles({
+				studio: getAllowStudio(),
+				configure: getAllowConfigure(),
+				developer: getAllowDeveloper(),
+				testing: getAllowTesting(),
+				service: getAllowService(),
+			})
+		} else if (user && subsReady) {
+			setRoles({
+				studio: getAllowStudio(),
+				configure: getAllowConfigure(),
+				developer: getAllowDeveloper(),
+				testing: getAllowTesting(),
+				service: getAllowService(),
+			})
+		}
+	}, [location.search, user, subsReady])
+
+	return roles
+}
+
+function useFeatureFlags() {
```

Comment:
Perhaps the return of this could be wrapped in a context which other places in the code can use to retrieve the values?
Same for `useRoles`.
That will allow us to start referencing this config in a reactive way, preparing for when we eventually want it to be fully reactive. This will be more important for the `useRoles`, as that depends on `user`

Reply:
I think it can, but I didn't want to inflate the scope of this PR too much. All I wanted to achieve here was to refactor these components as funcional components and modernize the Routes in the Router so that we can at some point do an easier switch to react-router-dom v6. Also, I'm not sure if we want to commit to using contexts, especially if we were to switch to a new state management, like MobX, or something. I think I'd rather keep it minimal for now, and do it when we make a decision on how to progress post-Meteor.